jetty-server/src/main/java/org/eclipse/jetty/server/ConnectionFactory.java

```diff
@@ -85,4 +86,51 @@
          */
         Connection upgradeConnection(Connector connector, EndPoint endPoint, MetaData.Request upgradeRequest, HttpFields responseFields) throws BadMessageException;
     }
+
+    /**
+     * <p>Connections created by this factory MUST implement {@link Connection.UpgradeTo}.</p>
+     */
+    interface Detecting extends ConnectionFactory
+    {
+        /**
+         * The possible outcomes of the {@link #detect(ByteBuffer)} method.
+         */
+        enum Detection
+        {
+            /**
+             * A {@link Detecting} can work with the given bytes.
+             */
+            RECOGNIZED,
+            /**
+             * A {@link Detecting} cannot work with the given bytes.
+             */
+            NOT_RECOGNIZED,
+            /**
+             * A {@link Detecting} requires more bytes to make a decision.
+             */
+            NEED_MORE_BYTES
+        }
+
+        /**
+         * <p>Check the bytes in the given {@code buffer} to figure out if this {@link Detecting} instance
+         * can work with them or not.</p>
+         * <p>The {@code buffer} MUST be left untouched by this method: bytes MUST NOT be consumed and MUST NOT be modified.</p>
+         * @param buffer the buffer.
+         * @return One of:
+         * <ul>
+         * <li>{@link Detection#RECOGNIZED} if this {@link Detecting} instance can work with the bytes in the buffer</li>
+         * <li>{@link Detection#NOT_RECOGNIZED} if this {@link Detecting} instance cannot work with the bytes in the buffer</li>
+         * <li>{@link Detection#NEED_MORE_BYTES} if this {@link Detecting} instance requires more bytes to make a decision</li>
+         * </ul>
+         */
+        Detection detect(ByteBuffer buffer);
+
+        /**
+         * Has this instance been configured with a next protocol? A {@link DetectorConnectionFactory} won't allow
+         * {@link Detecting} instances that do not have a configured next protocol.
+         * @return true if a next protocol has been configured, false otherwise.
+         */
+        boolean isNextProtocolConfigured();
```

Comment:
It is necessary if we want to prevent -for instance- a `ProxyConnectionFactory` with no configured next protocol inside a `DetectorConnectionFactory`.

I agree that this method makes `ConnectionFactory.Detecting` a bit uglier than it could be, but OTOH this allows catching invalid configurations early and reporting a clear error message. Without this, we cannot detect if a `ProxyConnectionFactory` used in a `DetectorConnectionFactory` has a configured next protocol as we agreed in solution 1 to enforce.